### ISSUE TITLE
Can't register new hosts via pxe or api in working-1.6

### ISSUE THREAD

**Issue by darksidemilk** [FOGProject/fogproject#548]
**Describe the bug**

When attempting to add a new host to fog in 1.6 either through the api or pxe full host registration it fails

**To Reproduce**

Steps to reproduce the behavior for pxe:
1. Install or upgrade to fog 1.6
2. pxe boot to it with an unregistered host (i.e a new vm)
3. Attempt to do a full host registration and inventory and see the error that your hostname is invalid but this host name displayed isn't what you typed but rather a random string

Steps to reproduce the behavior for pxe:
1. Install or upgrade to fog 1.6
2. Submit an api call to create a new host
3. Get a 500 error in return instead of a success message

Example using fogapi powershell module

```
install-module fogapi
import-module fogapi
set-fogserversettings -fogapitoken 'server-api-token-here' -fogusertoken 'your-fog-user-token-here' -fogserver 'your-fog-server-hostname-here';
Add-FogHost -name test-host -macs "01:02:03:04:05:06" 
```

**Expected behavior**

The host should have been added or registered successfully

**Screenshots**

Error on api call
![image](https://user-images.githubusercontent.com/7743340/221218325-65eeeb1c-2cf1-47b2-b206-5882932fb302.png)

Error on full pxe registration

![image](https://user-images.githubusercontent.com/7743340/221220449-8b6ba208-d071-4365-a787-50baecba3703.png)

**Software (please complete the following information):**
 - FOG version 1.6
 - FOS kernel 5.15.19
 - OS: Cent OS Stream 9 (Also occurred on Cent OS 7)
 - PHP: Occurs with both 7.4 and 8.0.27

**Additional context**

It seems to be related to a call to the `ishostnamesafe` function in the [host class](https://github.com/FOGProject/fogproject/blob/working-1.6/packages/web/lib/fog/host.class.php).

```
public function isHostnameSafe($hostname = '')
    {
        if (empty($hostname)) {
            $hostname = $this->get('name');
        }
        $pattern = "/^[\w!@#$%^()\-'{}\.~]{1,15}$/";
        return (bool)preg_match($pattern, $hostname);
    }
 ```

if I change this function so it always returns true, then it gets past the error in the pxe host registration

```
public function isHostnameSafe($hostname = '')
    {
        return true;
        //if (empty($hostname)) {
           // $hostname = $this->get('name');
        //}
        //$pattern = "/^[\w!@#$%^()\-'{}\.~]{1,15}$/";
        //return (bool)preg_match($pattern, $hostname);
    }
```
But of course, that's not what we want, we want this check to exist. 
Also, it does something else weird, it gives it a different name than what I input.
i.e. I input `dr-wks-jjv5` as the hostname in the pxe regsitration with the ishostnamesafe set to always return true.
It looks fine in the pxe menu but then the hostname that gets registered is `ZHItd2tzLWpqdjUK`

![image](https://user-images.githubusercontent.com/7743340/221225438-78920801-6d79-4c00-9ec3-8a4af84eba56.png)

So something is happening with the input between the [registration class fullreg call](https://github.com/FOGProject/fogproject/blob/working-1.6/packages/web/lib/reg-task/registration.class.php#L148) and when the host is registered.

Changing the return to always true doesn't get past the problem in the api call, so something else is happening with that.

**Comment by darksidemilk** [comment]
This commit https://github.com/FOGProject/fogproject/commit/b8e076d373e352131176a5c5f39d9b5b48c3666b appears to have fix pxe registration.

However, in for 1.5.X we can create a fog host via the api with a json like

```
{
  "name": "someName",
  "macs": [
    "1A:2B:3C:4D:5E:6F"
  ],
  "modules": [
    "4",
    "7",
    "1",
    "3",
    "5",
    "8",
    "9",
    "13",
    "10",
    "6",
    "11",
    "2",
    "12"
  ]
}
```

Which is the json format created by the fogApi powershell module function [New-FogHost](https://fogapi.readthedocs.io/en/latest/commands/New-FogHost/)

In working-1.6 this json produces a 500 error

However, if you remove the `macs` array property from the json, it gives an `{ "error": "Already created" }` in the response.
But, it actually does create the host

![image](https://github.com/FOGProject/fogproject/assets/7743340/3c0b1c47-c5b1-4d82-8a04-41f056ff0102)

If you hover over that host with no mac (it won't open on click) you can view the url that includes the host id, or using the fogApi ps module you can find the host id in the full list of hosts with `get-foghosts | ? name -eq 'someName' | select id`

Then you can use the api to add the mac to the host as a primary mac
`Add-FogHostMac -hostID (get-foghosts | ? name -eq 'someName').id -primary -macAddress "1A:2B:3C:4D:5E:6F" -vb`

This function will given an error saying 404 not found at one point because it tries to get the host record before adding the mac, until the mac is added you can't get the record of the host, only view it in the full list of hosts, same as in the UI. But this does add the mac via the api and 
![image](https://github.com/FOGProject/fogproject/assets/7743340/0a91c4d1-55a1-4228-a805-426283246afc)

However, if you were to add another mac address to that host and mark it as primary too, i.e.
`Add-FogHostMac -hostID (get-foghosts | ? name -eq 'someName').id -primary -macAddress "1A:2B:3C:4D:5E:6H"`
it creates 2 entries for that host in the hosts view
![image](https://github.com/FOGProject/fogproject/assets/7743340/b8ab4a6b-dea3-4c8f-9493-93e7da2d4218)

If you click into either it goes to the same place and lists the more recently set as primary mac as the primary mac
![image](https://github.com/FOGProject/fogproject/assets/7743340/aaa4ba2a-dfd2-403f-abc8-6d8e8d2c2212)

If you click the 'Primary Mac' selection in the UI, it fixes the duplicated display.
The duplicated display doesn't happen if you add a non-primary mac via the api

A bit hinky, but hopefully that workaround might help @mastacontrola with debugging what's going on.

Short version

* Host registration now works in pxe but not via api
* You can create a host with no mac via the api, and then add the mac to it after and this does work
* If you add a new primary mac via the api, it makes the ui display that host twice until you re-set a primary mac in the ui

**Comment by Neustradamus** [comment]
To follow

**Comment by darksidemilk** [comment]
This has been resolved. I don't recall exactly which commit, it may have been this one https://github.com/FOGProject/fogproject/commit/b8e076d373e352131176a5c5f39d9b5b48c3666b

But as of the current version for sure (https://github.com/FOGProject/fogproject/commit/e6a9f582e7180441338765ebab9e6b5edbca81e4) this is resolved.
